### PR TITLE
nmcli.py by itself

### PR DESCRIPTION
**Pull request by alcamie101**
If you could review this thanks Will

### CODE REVIEW

File: library/nmcli.py [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+author: Chris Long
+version_added: "0.1"
+short_description: Manage Networking
+requirements: [ conadd, condel, conmod, conup, condown, constatus, conshow ]
+description:
+    - Manage the network devices. Create, modify, and manage, ethernet, teams, bonds, vlans etc.
+options:
+    name:
+        required: true
+        version_added: "0.1"
```
Comment (willthames) [inline library/nmcli.py]: version_added relates to Ansible version. And only really needed if it differs to the version_added of the module itself.

File: library/nmcli.py [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+# along with Ansible.    If not, see <http://www.gnu.org/licenses/>.
+
+
+DOCUMENTATION='''
+---
+module: nmcli
+author: Chris Long
+version_added: "0.1"
+short_description: Manage Networking
+requirements: [ conadd, condel, conmod, conup, condown, constatus, conshow ]
```
Comment (willthames) [inline library/nmcli.py]: what are these?

File: library/nmcli.py [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+# To lock a profile to a specific interface, issue a command as follows
+- nmcli: name=add ctype=ethernet name=my-eth1 ifname=eth1 state=present
+
+# To make a profile usable for all compatible Ethernet interfaces, issue a command as follows
+- nmcli: name=add ctype=ethernet name=my-eth1 ifname="*" state=present
+
+# To change the property of a setting e.g. MTU, issue a command as follows:
+- nmcli: name=modify cname=my-eth1 mtu=9000
+
+    Exit Status's:
```
Comment (willthames) [inline library/nmcli.py:246]: Do we care about the Exit Statuses here?

File: library/nmcli.py [hunk 4]
```diff
@@ -0,0 +1,735 @@
...
+        self.priority=module.params['priority']
+        self.slavepriority=module.params['slavepriority']
+        self.forwarddelay=module.params['forwarddelay']
+        self.hellotime=module.params['hellotime']
+        self.maxage=module.params['maxage']
+        self.ageingtime=module.params['ageingtime']
+        self.mac=module.params['mac']
+        self.vlanid=module.params['vlanid']
+        # The following is going to be used in dbus code
+        self.devtypes={1: "Ethernet",
```
Comment (willthames) [inline library/nmcli.py]: I'd make a constant called DEVTYPES and refer to that here.

File: library/nmcli.py [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+                       8: "Modem",
+                       9: "InfiniBand",
+                       10: "Bond",
+                       11: "VLAN",
+                       12: "ADSL",
+                       13: "Bridge",
+                       14: "Generic",
+                       15: "Team"
+                       }
+        self.states={0: "Unknown",
```
Comment (willthames) [inline library/nmcli.py]: same for constant dict STATES

File: library/nmcli.py [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+            # except KeyError:
+            #     state = "Unknown"
+            # print "State: %s" % state
+        return connection_list
+
+    def connection_exists(self):
+        # we are going to use name and type in this instance to find if that connection exists and is of type x
+        connections=self.list_connection_info()
+
+        for con_item in connections:
```
Comment (willthames) [inline library/nmcli.py:446]: return con_item in connections

File: library/nmcli.py [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+            maxage=dict(required=False, default="20", type='str'),
+            ageingtime=dict(required=False, default="300", type='str'),
+            # vlan specific vars
+            vlanid=dict(required=False, default=None, type='str'),
+            # following options are specific to ifnamedel
+        ),
+        supports_check_mode=True
+    )
+
+    # if module.params['state'] is None and module.params['enabled'] is None:
```
Comment (willthames) [inline library/nmcli.py]: delete these, and set them to required=True

File: library/nmcli.py [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+
+        # elif not nmcli.connection_exists():
+        #     result['Connection']='No Connection named %s exists' % nmcli.cname
+        #     (rc, out, err)=nmcli.connection_down()
+        #     (rc, out, err)=nmcli.remove_connection()
+        #     result['Connection']='Connection %s is being removed' % nmcli.cname
+        #     result['Connection']='Yessss. Connection %s exists' % nmcli.cname
+        #     result['Connection']='Connection %s is being set to down' % nmcli.cname
+        if rc!=0:
+            module.fail_json(name =('No Connection named %s exists' % nmcli.cname), msg=err, rc=rc)
```
Comment (willthames) [inline library/nmcli.py:889]: Don't fail when a connection that is specified as absent is already absent. exit with changed=False.

File: library/nmcli.py [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+
+    rc=None
+    out=''
+    err=''
+    result={}
+    result['cname']=nmcli.cname
+    result['state']=nmcli.state
+
+    if nmcli.state=='absent':
+        # if nmcli.connection_exists():
```
Comment (willthames) [inline library/nmcli.py]: Does connection_exists() work - I'd think this should be the priority,

File: library/nmcli.py [hunk 10]
```diff
@@ -0,0 +1,735 @@
...
+
+        cmd=[self.module.get_bin_path('nmcli', True)]
+        cmd.append('con')
+        cmd.append('del')
+        cmd.append(self.cname)
+        return self.execute_command(cmd)
+
+    def modify_connection(self):
+        cmd=[]
+        if self.type=='team':
```
Comment (willthames) [inline library/nmcli.py:794]: you can create a dict of type to method call

commands = { 'team': self.modify_connection_team, 'team-slave', ... }
return self.execute_command(commands[self.type])

File: library/nmcli.py [hunk 11]
```diff
@@ -0,0 +1,735 @@
...
+            cmd=self.modify_connection_vlan()
+        return self.execute_command(cmd)
+
+
+def main():
+    # Parsing argument file
+    module=AnsibleModule(
+        argument_spec=dict(
+            enabled=dict(required=False, default=None, choices=['yes', 'no'], type='str'),
+            name=dict(required=False, default=None, choices=['add', 'mod', 'show', 'up', 'down', 'del'], type='str'),
```
Comment (willthames) [inline library/nmcli.py]: this is a bad field name. Use 'action' or similar. Also, suggest having nice option names like 'modify' and 'delete' and converting while required. 

Actually, this field shouldn't even exist. Use enabled and state to cover these cases (you can add 'show' to state)

File: library/nmcli.py [hunk 12]
```diff
@@ -0,0 +1,735 @@
...
+            state=dict(default=None, choices=['present', 'absent', 'modify'], type='str'),
```
Comment (willthames) [inline library/nmcli.py]: modify is not a state. Use present, and check if the parameters are the same, modifying if necessary

File: library/nmcli.py [hunk 13]
```diff
@@ -0,0 +1,735 @@
...
+        # if connection_exists()
+        #     self.module.fail_json(msg="this connection exists already")
+        # else:
+        # format for creating team interface
+        cmd.append('connection')
+        cmd.append('add')
+        cmd.append('type')
+        cmd.append(self.type)
+        cmd.append('con-name')
+        if self.cname is not None:
```
Comment (willthames) [inline library/nmcli.py:473]: this whole next section should just be replaced with some kind of for loop

```
for attr in ['cname', 'ip4, 'gw4', ... ]: 
    if getattr(self, attr) is not None: 
        cmd.append(attr)
        cmd.append(getattr(self, attr))
```

File: library/nmcli.py [hunk 14]
```diff
@@ -0,0 +1,735 @@
...
+            cmd.append(self.cname)
+        elif self.ifname is not None:
+            cmd.append(self.ifname)
+        else:
+            self.module.fail_json(msg="You haven't specified a name for the connection")
```
Comment (willthames) [inline library/nmcli.py]: Handle this case long before it gets to this method call. This method shouldn't need to access module at all.

File: library/nmcli.py (hunk 2 above)
Comment (alcamie101) [inline library/nmcli.py]: thanks for pointing that out. Should be just NMCLI at present

File: library/nmcli.py (hunk 3 above)
Comment (alcamie101) [inline library/nmcli.py:246]: Just notes to assist with debugging.

File: library/nmcli.py (hunk 9 above)
Comment (alcamie101) [inline library/nmcli.py]: I was getting some weird happenings the other day. connection_exists() does now work. Will put that test back in.

File: library/nmcli.py (hunk 12 above)
Comment (alcamie101) [inline library/nmcli.py]: Yeh. Cutting down the number of fields required is definitely a good thing

File: library/nmcli.py (hunk 13 above)
Comment (alcamie101) [inline library/nmcli.py:473]: OK I'll try that.

File: library/nmcli.py (hunk 14 above)
Comment (alcamie101) [inline library/nmcli.py]: Yeh, that way there are a lot less fail_json's to write

**Pull request merged**